lib/ex_doc/refs.ex

```diff
@@ -132,8 +135,26 @@ defmodule ExDoc.Refs do
     end
   end
 
-  defp visibility(:hidden), do: :hidden
-  defp visibility(_), do: :public
+  defguardp is_kind(kind) when kind in [:callback, :function, :module, :type]
+  defguardp no_docs?(doc) when doc == :none or (is_map(doc) and map_size(doc) == 0)
+
+  defp visibility(kind, _name, :hidden) when is_kind(kind),
+    do: :hidden
+
+  defp visibility(kind, _name, doc) when kind in [:callback, :module, :type] and no_docs?(doc),
+    do: :public
+
+  defp visibility(:function, name, doc) when no_docs?(doc) do
+    if hd(Atom.to_charlist(name)) == ?_ do
+      :hidden
+    else
+      :public
+    end
+  end
+
+  defp visibility(kind, _name, _doc) when is_kind(kind) do
+    :public
+  end
```

Comment:
Is there any case where we would want to behave differently than `ExDoc.Retriever.doc?`? If that's not the case, then perhaps we could extract that boolean `doc?` function and then just convert it to `:public/:hidden` here?

Reply:
This is the only suggestion that I have not implemented. Do you mind sharing the implementation of what you mean?

Reply:
I looked into that and the intent of these functions is in fact different - `ExDoc.Retriever.doc?` is used only to filter which functions/macros should be shown in the doc.

I do however think it would make sense to have a single source of truth regarding which elements get documentation and which are ignored. Currently there are two separate sources:

- the `Retriever` which ignores a given module/function/macro/... based on some rules
- the `Refs` cache that keeps track of module/function/macro/... "visibility"

The question is whether these two should be in line with each other in all cases. The issue hints me that they should, but I may be missing something here.

What I had in mind is a single function that would determine if the given module or function should end up in the generated documentation or rather be ignored. This function could then be used in all the relevant places. [Here](https://github.com/eksperimental/ex_doc/compare/functions_starting_with_underscore...jonatanklosko:pr1313) I implemented such function (`should_have_doc?`) and tried to make it match the current behavior I saw in several different places. I think it removes much redundancy, but someone should confirm that this approach makes sense.

Reply:
@jonatanklosko  I think covering those issues in this PR it will deviate the purpose too much.
I have taken the ideas of unifying regarding to visibility and put it all in one function.
So now what this PR does is laying the foundation for the logic regarding the visibility of functions.
I think once we merge this we can look into your proposal. What do you think?

Thank you for your contributions

Reply:
Sure, what I suggested was more of a refactoring and can definitely be discussed separately. Your changes surely fix the problem.